File: sknetwork/data/test_graphs.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Nov 29, 2018
@author: Quentin Lutz <qlutz@enst.fr>
@author: Nathan de Lara <nathan.delara@polytechnique.org>
@author: Thomas Bonald <tbonald@enst.fr>
"""

import numpy as np
from scipy import sparse
import pytest

from sknetwork.utils import directed2undirected


@pytest.mark.skip
def test_digraph() -> sparse.csr_matrix:
    """Simple directed graph, used for testing.
    10 nodes, 12 edges
    """
    row = np.array([0, 1, 1, 3, 4, 6, 6, 6, 6, 7, 8, 8, 9])
    col = np.array([1, 4, 3, 2, 5, 4, 5, 7, 1, 9, 9, 2, 9])
    data = np.array([1, 1, 2.5, 1, 2, 2, 1, 2, 2, 1.5, 2, 1, 2])
    return sparse.csr_matrix((data, (row, col)), shape=(10, 10))


@pytest.mark.skip
def test_graph() -> sparse.csr_matrix:
    """Simple undirected graph, used for testing.
    10 nodes, 12 edges.
    """
    return directed2undirected(test_digraph(), weighted=True)


@pytest.mark.skip
def test_bigraph() -> sparse.csr_matrix:
    """Simple bipartite graph, used for testing.
    6 + 8 nodes, 9 edges.
    """
    row = np.array([0, 1, 1, 2, 2, 3, 4, 5, 5])
    col = np.array([1, 2, 3, 1, 0, 4, 7, 5, 6])
    data = np.array([1, 2.5, 1, 2, 2, 1.5, 1, 2, 3])
    return sparse.csr_matrix((data, (row, col)), shape=(6, 8))


@pytest.mark.skip
def test_disconnected_graph() -> sparse.csr_matrix:
    """Simple disconnected undirected graph, used for testing.
    10 nodes, 10 edges.
    """
    row = np.array([1, 2, 3, 4, 6, 6, 6, 7, 8, 9])
    col = np.array([1, 3, 2, 5, 4, 5, 7, 9, 9, 9])
    data = np.array([1, 2.5, 1, 2, 2, 1, 2, 2, 1.5, 2])
    adjacency = sparse.csr_matrix((data, (row, col)), shape=(10, 10))
    return directed2undirected(adjacency)


@pytest.mark.skip
def test_bigraph_disconnect() -> sparse.csr_matrix:
    """Simple disconnected bipartite graph, used for testing.
    6 + 8 nodes, 9 edges.
    """
    row = np.array([1, 1, 1, 2, 2, 3, 5, 4, 5])
    col = np.array([1, 2, 3, 1, 3, 4, 7, 7, 6])
    data = np.array([1, 2.5, 1, 2, 2, 1.5, 3, 0, 1])
    return sparse.csr_matrix((data, (row, col)), shape=(6, 8))


@pytest.mark.skip
def test_graph_bool() -> sparse.csr_matrix:
    """Simple undirected graph with boolean entries, used for testing (10 nodes, 10 edges)."""
    adjacency = test_graph()
    adjacency.data = adjacency.data.astype(bool)
    return adjacency


@pytest.mark.skip
def test_clique() -> sparse.csr_matrix:
    """Clique graph, used for testing (10 nodes, 45 edges).
    """
    n = 10
    adjacency = sparse.csr_matrix(np.ones((n, n), dtype=bool))
    adjacency.setdiag(0)
    adjacency.eliminate_zeros()
    return adjacency


@pytest.mark.skip
def test_graph_empty() -> sparse.csr_matrix:
    """Empty graph, used for testing (10 nodes, 0 edges).
    """
    return sparse.csr_matrix((10, 10), dtype=bool)

File: sknetwork-rs/src/data/test_graphs.rs
use sprs::{CsMat, TriMat};

use crate::utils::format::directed2undirected;

/// Returns a 10-node weighted directed test graph.
pub fn test_digraph() -> CsMat<f64> {
    let row = [0usize, 1, 1, 3, 4, 6, 6, 6, 6, 7, 8, 8, 9];
    let col = [1usize, 4, 3, 2, 5, 4, 5, 7, 1, 9, 9, 2, 9];
    let data = [
        1.0f64, 1.0, 2.5, 1.0, 2.0, 2.0, 1.0, 2.0, 2.0, 1.5, 2.0, 1.0, 2.0,
    ];
    let mut tri = TriMat::<f64>::new((10, 10));
    for i in 0..row.len() {
        tri.add_triplet(row[i], col[i], data[i]);
    }
    tri.to_csr::<usize>()
}

/// Returns the symmetrized version of [`test_digraph`].
pub fn test_graph() -> CsMat<f64> {
    directed2undirected(&test_digraph(), true)
}

/// Returns a 6×8 weighted bipartite test graph.
pub fn test_bigraph() -> CsMat<f64> {
    let row = [0usize, 1, 1, 2, 2, 3, 4, 5, 5];
    let col = [1usize, 2, 3, 1, 0, 4, 7, 5, 6];
    let data = [1.0f64, 2.5, 1.0, 2.0, 2.0, 1.5, 1.0, 2.0, 3.0];
    let mut tri = TriMat::<f64>::new((6, 8));
    for i in 0..row.len() {
        tri.add_triplet(row[i], col[i], data[i]);
    }
    tri.to_csr::<usize>()
}

/// Returns a 10-node undirected graph with an isolated component.
pub fn test_disconnected_graph() -> CsMat<f64> {
    let row = [1usize, 2, 3, 4, 6, 6, 6, 7, 8, 9];
    let col = [1usize, 3, 2, 5, 4, 5, 7, 9, 9, 9];
    let data = [1.0f64, 2.5, 1.0, 2.0, 2.0, 1.0, 2.0, 2.0, 1.5, 2.0];
    let mut tri = TriMat::<f64>::new((10, 10));
    for i in 0..row.len() {
        tri.add_triplet(row[i], col[i], data[i]);
    }
    directed2undirected(&tri.to_csr::<usize>(), true)
}

/// Returns a bipartite graph with a disconnected row block.
pub fn test_bigraph_disconnect() -> CsMat<f64> {
    let row = [1usize, 1, 1, 2, 2, 3, 5, 4, 5];
    let col = [1usize, 2, 3, 1, 3, 4, 7, 7, 6];
    let data = [1.0f64, 2.5, 1.0, 2.0, 2.0, 1.5, 3.0, 0.0, 1.0];
    let mut tri = TriMat::<f64>::new((6, 8));
    for i in 0..row.len() {
        tri.add_triplet(row[i], col[i], data[i]);
    }
    tri.to_csr::<usize>()
}

/// Returns [`test_graph`] with all nonzero weights set to `1.0`.
pub fn test_graph_bool() -> CsMat<f64> {
    let mut g = test_graph();
    for x in g.data_mut() {
        *x = if *x != 0.0 { 1.0 } else { 0.0 };
    }
    g
}

/// Returns a 10-node unweighted clique adjacency matrix.
pub fn test_clique() -> CsMat<f64> {
    let n = 10usize;
    let mut tri = TriMat::<f64>::new((n, n));
    for i in 0..n {
        for j in 0..n {
            if i != j {
                tri.add_triplet(i, j, 1.0);
            }
        }
    }
    tri.to_csr::<usize>()
}

/// Returns a 10×10 zero adjacency matrix.
pub fn test_graph_empty() -> CsMat<f64> {
    CsMat::<f64>::zero((10, 10))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_shapes() {
        assert_eq!(test_digraph().shape(), (10, 10));
        assert_eq!(test_graph().shape(), (10, 10));
        assert_eq!(test_bigraph().shape(), (6, 8));
        assert_eq!(test_disconnected_graph().shape(), (10, 10));
        assert_eq!(test_bigraph_disconnect().shape(), (6, 8));
        assert_eq!(test_clique().shape(), (10, 10));
        assert_eq!(test_graph_empty().shape(), (10, 10));
    }
}
